pkg/phlaredb/compact.go

```diff
@@ -534,3 +473,463 @@ func prepareIndexWriter(ctx context.Context, path string, readers []BlockReader)
 
 	return indexw, nil
 }
+
+type symbolsRewriter struct {
+	profiles    iter.Iterator[profileRow]
+	rewriters   map[BlockReader]*stacktraceRewriter
+	stacktraces []uint32
+	err         error
+
+	numSamples uint64
+}
+
+func newSymbolsRewriter(it iter.Iterator[profileRow], blocks []BlockReader, w SymbolsWriter) *symbolsRewriter {
+	sr := symbolsRewriter{
+		profiles:  it,
+		rewriters: make(map[BlockReader]*stacktraceRewriter, len(blocks)),
+	}
+	for _, r := range blocks {
+		sr.rewriters[r] = newStacktraceRewriter(r.Symbols(), w)
+	}
+	return &sr
+}
+
+func (s *symbolsRewriter) NumSamples() uint64 { return s.numSamples }
+
+func (s *symbolsRewriter) At() profileRow { return s.profiles.At() }
+
+func (s *symbolsRewriter) Close() error { return s.profiles.Close() }
+
+func (s *symbolsRewriter) Err() error {
+	if s.err != nil {
+		return s.err
+	}
+	return s.profiles.Err()
+}
+
+func (s *symbolsRewriter) Next() bool {
+	if !s.profiles.Next() {
+		return false
+	}
+	var err error
+	profile := s.profiles.At()
+	profile.row.ForStacktraceIDsValues(func(values []parquet.Value) {
+		s.loadStacktracesID(values)
+		r := s.rewriters[profile.blockReader]
+		if err = r.rewriteStacktraces(profile.row.StacktracePartitionID(), s.stacktraces); err != nil {
+			return
+		}
+		s.numSamples += uint64(len(values))
+		for i, v := range values {
+			// FIXME: the original order is not preserved, which will affect encoding.
+			values[i] = parquet.Int64Value(int64(s.stacktraces[i])).Level(v.RepetitionLevel(), v.DefinitionLevel(), v.Column())
+		}
+	})
```

Comment:
When we translate stack trace IDs, the original order will be invalidated:

Before compaction:
```
IDs: 1, 2, 3, 4, 42, 43
```
After compaction:
```
IDs: 1, 2, 6, 4, 453, 12
```

It will deteriorate delta encoding compression ratio. We clearly should fix this